### DISCUSSION TITLE
Possible improvements to the build process

### DISCUSSION THREAD

**Discussion by cynova** [directus/directus#6633] [Feature Requests]
There are three issues here from my POV:

1. There is no obvious way to build Directus docker Images yourself. The official process is convoluted, hacky and just plain weird (publish to NPM and then install that?!). Sure, one can write their own scripts (which is what we did), but then we have no guarantee that our build is as close as possible to the official one (except for the code changes of course).
2. The current official image reinstalls Directus *on every restart*. WTF? Directus is insanely fast - but now we deal with boottimes that are completely nuts, especially for extension development, because there is no way of reloading extensions.
3. The entrypoint file. Why is this logic not in Node? When running directus without Docker there must be some other way of for example creating the sqlite folder, so why not just use that?

I really don't see what is preventing a simple, straightforward build process without a lot of magic.
I just want to know whether i misinterpreted something or if there is some official way of solving above problems before i go back to writing even more hacks to solve them for us only :D

**Comment by rijkvanzanten** [comment]
Feel free to open a PR for that flow with any and all improvements that you can think of 👍🏻 

PS, it's way more helpful if you frame these sorts of things as "Hey, I found a couple things we can optimize" instead of "Hey, why is this so bad" 

As for your actual questions:

> (publish to NPM and then install that?!)

How else would you propose doing it? The Directus release on NPM is the actual distribution that we want/need people to use

> The current official image reinstalls Directus on every restart

That sounds like something we could optimize

> The entrypoint file. Why is this logic not in Node?

That to me sounds like an overcomplication 🤷🏻 Trying to use node to execute another node process sounds like a problem waiting to happen. Doing that in bash makes more sense to me

**Reply by cynova** [reply]
> Feel free to open a PR for that flow with any and all improvements that you can think of 👍🏻
OK, I'll try to do that this or next week.

> PS, it's way more helpful if you frame these sorts of things as "Hey, I found a couple things we can optimize" instead of "Hey, why is this so bad"

I know, i wasn't trying to sound condescending. I'm just genuinely confused because Directus has better features compared to the competition, is developed at a rapid pace, has amazing code quality - and the weirdest pipeline i've ever seen. This is not meant as an insult, i'm just trying to figure out if there is an underlying reason before i try to fix something that maybe can't be fixed (and is known to be unfixable.)

> As for your actual questions:
> 
> > (publish to NPM and then install that?!)
> 
> How else would you propose doing it? The Directus release on NPM is the actual distribution that we want/need people to use
I'd propose to build the package and then pass the output to two jobs; one to publish to NPM and one to package as docker container.

I'm not 100% sure but i think the current solution doesn't even lock the directus version, so subsequent restarts of a directus @rc-82 image would download @rc-83 for example. (really not sure here though, didn't actually test it). If this were the case it kind of defeats one of the purposes of using docker (immutability).

> > The current official image reinstalls Directus on every restart
> 
> That sounds like something we could optimize
If we do as proposed above this problem will go away and boot times should be back to the ~2 seconds they were a few versions ago.

> > The entrypoint file. Why is this logic not in Node?
> 
> That to me sounds like an overcomplication 🤷🏻 Trying to use node to execute another node process sounds like a problem waiting to happen. Doing that in bash makes more sense to me

I agree on the overcomplication argument, but the question i'm asking is more why the entrypoint file is even necessary - locally installed directus via npm/npx doesn't need it either. So either there is some manual intervention needed  (creating sqlite directory) in case of the local install (which is probably not what should happen anyways) - or there is something else at work there that i didn't see.
I'll try to figure this out asap.

**Reply by rijkvanzanten** [reply]
I don't know about the version (cc @paescuj / @t7tran). The entry point runs directus bootstrap and start, to bootstrap / migrate the database before starting the server to ensure updates between versions are handled properly.

**Reply by cynova** [reply]
I'm trying to work on this but still have a hard time figuring out how everything comes together.
Do you publish to NPM manually via lerna (npm run release), before publishing on github?
I saw create-release.yml, but that seems to only create a github release - at least i can't see how it's supposed to publish to NPM.

**Reply by paescuj** [reply]
> I'm not 100% sure but i think the current solution doesn't even lock the directus version, so subsequent restarts of a directus @rc-82 image would download @rc-83 for example. (really not sure here though, didn't actually test it). If this were the case it kind of defeats one of the purposes of using docker (immutability).

The installation of the `directus` package has failed during the image [build process in v9.0.0-rc.83](https://github.com/directus/directus/runs/2946161276). Because of this, the  package is now being installed on container startup. This has been addressed in #6601 and should work again as expected once v9.0.0-rc.84 is released. (@rijkvanzanten In the meantime, you may want to manually trigger a rebuild of the image for v9.0.0-rc.83 to fix this).
Besides of the case described above, the installation of the `directus` package takes place during the build process and thus the image is shipped with the correct version. Additionally, I just submitted a small PR (#6659) to ensure we still end up with the desired version, if a newer version of `directus` has been published in the meantime.

> I agree on the overcomplication argument, but the question i'm asking is more why the entrypoint file is even necessary - locally installed directus via npm/npx doesn't need it either. So either there is some manual intervention needed (creating sqlite directory) in case of the local install (which is probably not what should happen anyways) - or there is something else at work there that i didn't see.
> I'll try to figure this out asap.

As @rijkvanzanten already pointed out, the entrypoint script is used to bootstrap / migrate the Directus instance. This makes perfect sense, IMO. Using an entrypoint script to bootstrap / initialize a Docker container is a common pattern.

> > (publish to NPM and then install that?!)
> 
> How else would you propose doing it? The Directus release on NPM is the actual distribution that we want/need people to use

I was wondering the same thing. The current process makes sense to me, although I agree that we could improve the sequence here -> Wait with the `build-image` action until the `create-release` action (to publish the NPM packages?) has been completed successfully.

> There is no obvious way to build Directus docker Images yourself

You can try to build it yourself by using the following Makefile: https://github.com/directus/directus/blob/b4dd1993da46bed95fa582d7bff0c11f31204871/.github/actions/Makefile
However, I agree that there is room for improvment (documentation, etc.).

Generally, I think the Docker image is quite good. Just a few weeks ago @t7tran and me did some enhancements (reduce layers/size, use non-root user, ...). I agree, that the pipeline/action could be improved a bit and I'm glad to see your suggestions here.

**Reply by paescuj** [reply]
> I'm trying to work on this but still have a hard time figuring out how everything comes together.
> Do you publish to NPM manually via lerna (npm run release), before publishing on github?
> I saw create-release.yml, but that seems to only create a github release - at least i can't see how it's supposed to publish to NPM.

If you can be a little more specific about what you want to work on, I might be able to help you out.
However, I'm not sure about NPM publish part. I guess this is currently done manually.

**Reply by cynova** [reply]
@paescuj great!
OK, the part about rc83 explains the reinstalling, and the version pinning. so that was only bad luck.
So on Directus without docker migrating has to be done manually? Is there a reason Directus doesn't do that automatically on startup?

What i wanted to work on (if you agree that this approach would be better than the current one):

on tag push:
stage 1, job 1: npm run build
stage 2 (parallelised), job 1: copy prebuilt directus, then npm run release (what rjik likely does manually now)
stage 2 (parallelised), job 2: copy prebuilt directus, then docker build

I see the following advantages:
- Faster publishing
- No risk of what happened in rc83
- Container building + NPM publishing should work out of the box for forks if they set their own dockerhub/npm tokens (i would have to document this of course).

**Reply by paescuj** [reply]
> OK, the part about rc83 explains the reinstalling, and the version pinning. so that was only bad luck.
> So on Directus without docker migrating has to be done manually? Is there a reason Directus doesn't do that automatically on startup?

Bootstrapping is part of the [create-directus-project](https://docs.directus.io/guides/installation/cli/#_2-create-a-project) utility or can be done manually via [CLI](https://docs.directus.io/reference/command-line-interface/#initialize-a-new-project), migrations after an upgrade needs to be done manually (see [Upgrading / Updating a Project](https://docs.directus.io/guides/projects/#upgrading-updating-a-project)).
It might be possible to automate the migration stuff (@rijkvanzanten what's your opinion here?).

> on tag push:
> stage 1, job 1: npm run build
> stage 2 (parallelised), job 1: copy prebuilt directus, then npm run release (what rjik likely does manually now)
> stage 2 (parallelised), job 2: copy prebuilt directus, then docker build

This sounds good to me so far! Although I think you cannot simply copy the Directus build to the Docker image, right? You would need to do some cleanup, etc... Have you already thought about this?

**Reply by t7tran** [reply]
Well, the version thingy was just a bad side effect of the bug in the build process which surfaced in rc.83 and definitely needs to be improved, i.e. a more reliable way to proceed if package has been published. Other than that, it isn't a problem at all as the version is already included in the Dockerfile's install command. The recent PR #6659 is just a reassurance when this bug happens, no real value.

About the comment on the build process, it was the way it's supposed to be. I build lots of Docker images and still yet see a major problem with Directus's Dockerfile (@paescuj has done a good job on the recent cleanup). Docker/Containerisation helps simplifying the process, it doesn't make IT complexity disappear but concealing it. If one uses Docker and would like to have a personalised Directus image (or any products out there offering a Docker image), there are some options:

1. Build it from the scratch and consider contribute it back to the community
2. Comprehend and suggest enhancements/PR to the existing process to support their use cases
3. Extends the base image with customisations

The last option is always my recommendation. This takes away lots of maintenance work that one wouldn't understand while focusing on the customisation bit which they have total control, i.e. switch back to `root` to install more packages if required.

I haven't seen a well architectured/written project like Directus. The team offers a free product (where you can literally scratch their name off the product), proactive and friendly support forum, patience in their responses (pardon my repetitive suggestions), etc. So while there are gaps in addressing our needs, I always try to be on the constructive side.

**Reply by SeanDylanGoff** [reply]
@t7tran i agree generally but what i see different (and tell me if it's wrong):
I'm not talking about the dockerfile itself, i'm talking about the *process* of publishing. Publishing manually to npm first and then pulling from there might seem like a good idea generally but has a fundamental flaw: it is neither fullly automated nor standalone.
A fork CANNOT build the container without connecting their own npm account and publishing as a public package - and even then they have to replace a hardcoded package name anyways.

It's not impossible nor really complicated to do so, but it just seems less than optimal to me. Because that means a fork will always have to maintain an at least a bit different pipeline config/dockerfile for no reason.

Let me be really clear here, i do not believe it's the maintainers responsibility to handle this use case. They give us an amazing headless CMS for free and as you said - even with full whitelabeling support. I was trying to figure out why they didn't do it as i proposed - if my proposal is maybe flawed in some way i didn't see. i asked this because every bit of code in directus seems to had a lot of thought put into it - so i expected a big reason.

Now that i think i understand the build process as a whole i can work on a PR implementing it as described in my last comment.

**Reply by rijkvanzanten** [reply]
The main reasons are ease-of-installation and ease-of-maintenance. Being able to install Directus with a single "npm init create-directus-project" and/or "npm install directus" is invaluable for a project like this. It also means that we ensure that everybody is using the exact same build of Directus. If we'd end up having people build from source for production use (even when in Docker), there's always a chance that the build artifacts are ever so slightly different, based on versions of node and devdependencies (among other quirks, like installation differences between yarn and npm etc).

Secondly, from a maintenance perspective, we rely on npm workspaces to keep the codebase organized, which also means we rely on multiple npm packages tied together to form the final product. This makes long-term maintenance ~~way easier~~ possible. You could theoretically run the GitHub repo from source by building it, and running the `api` folder as you would with the production release, but as you already mentioned above, i do not believe it's the maintainers responsibility to handle this use case.

**Comment by br41nslug** [comment]
Heya! Thank you for taking the time to submit this request! 

It has been over **90 days**, and this discussion has not received at least **15 votes** from the community. This means that we don't feel like there's enough community interest to warrant further R&D into this topic at this time. :ice_cube:

This request will now be closed to keep our discussions tidy. Please reach out if you have any questions!

_For more information, see our [Feature Request Process](https://docs.directus.io/contributing/feature-request-process.html)._
